### PR TITLE
Adding SonarQube-ecoCode container image

### PR DESCRIPTION
**Pull request by obeone**
It use a multistage Dockerfile, so a maven container is used to build jar files, and they are just copied in SonarQube image, without building tools.

I updated documentation to add the docker command on main README (to allow quick test), and in INSTALL documentation, I divided it in three parts : quick start (test version), production version (dockerfile, with this image + postgresql) and a development version, with detailed build instructions and docker-compose start (almost all the original instruction).

I pushed a manual build on my package repos but it can be automatised

### CODE REVIEW

**Comment by dedece35** [comment]
hello @obeone ,
it's a good idea .. but before checking your PR, could you correct the previous build check ?
thanks.

**Comment by obeone** [comment]
Yes, I will :) 

I was waiting for some interest before dealing with merging latest commits. 

I work on it quickly

**Comment by dedece35** [comment]
hello @obeone,

it's a good idea to improve docker installation system. But there is another PR with the same kind of idea : see #89 
Maybe merge these two ideas into another new PR in order to improve our docker installation system.
(please see PR #89 created by @Silicoman / @MP-Aubay and my comment)

My feedbacks of this PR : 
- I don't known github workflow system, I can't tell you if ok for me or not ... @MP-Aubay ? @glalloue ? @jules-delecour-dav ? @jhertout ?
- maybe put sonarQube version on pom.xml file as the other PR

**Comment by obeone** [comment]
Hi @dedece35,

I answered on PR #89.

GitHub Workflow is a CI/CD tool (GitHu giving free compute time), so there is no need to have someone who build container, push it etc... It's automatic, and allow an always up-to-date container, built only once for everyone

**Comment by dedece35** [comment]
@obeone, 
thank you for answers ! very interesting !
to give only one place for answers, I answer you in your PR :p 

I agree with to have an ecocode plugin version deployed in github docker repository. Thus, everybody can use an already deployed version without to build it locally.
But in other PR #89, I find interesting that the version of plugin deployed is dynamic and managed from pom.xml. In your solution, I think, the lack is that the version of plugin is always 1.0.0-SNAPSHOT in `Dockerfile`. Isn't it ? but maybe I miss something ?
(same answer for sonarqube version managment because it is present on `Dockerfile` and `docker-compose.dev.yml`)
Actually, the process will always deploy the version 1.0.0-SNAPSHOT and replace the old one, isn't it ?
If yes, (maybe in other PR), we can do modifications to automatically replace "SNAPSHOT" suffix with a new one like date in "YYYYMMDDHHmmss", no ?

In conclusion, your solution is ready to be merged. Thus, I agree to merge this PR, and then make improvements to manage dynamically plugin versions and sonarqube versions.
What do you think about it @MP-Aubay, @glalloue, @jules-delecour-dav, @jhertout, @Silicoman ?

**Comment by obeone** [comment]
Sorry if my message could have seemed condescending, it was not the idea at all! 😬

Here is my limit : I don't do Java. 
I wrote the first version of this Dockerfile just before the ecoCode challenge, because I don't want to install jdk etc on my laptop. 

Clearly we can do better, in particular for the plugins version. 

For SonarQube version, plug-ins in June wasn't compatible with latest version. 

Since then, I have clearly made progress on Dockerfiles. If a Java developer is motivated to think together a little about doing better, it's with pleasure!

**Comment by Silicoman** [comment]
Hi everybody,

I'm not sure of the implementation direction.

_**First question, who is the target of this plugin ?**_
Users who actually are already using sonarqube and want to improve their instance with a new plugin. Users can use a variety of sonarqube version (v8 LTS (will soon go to v9), v9+, community vs enterprise). They can also build their own image sonarqube with their collection plugins and adding RUN step with "wget github.com/releases/xxx". So you won't distribute docker image but JAR files.

Another risk which isn't analyse, is the question of licence. You are distributing an image from sonarsource with a new licence GPLv3 due to contamination. Previous analyse wasn't check this aspect.

However, +1 about multistage Dockerfile. It's a good idea !

Best regards,

**Comment by dedece35** [comment]
Hi @obeone, @Silicoman,

First of all, thank you for interesting answers.

@obeone, there is no problem with "condescending" message. I don't understand it like this.
I only want this docker idea to progress.
I understand your purpose to have a docker image to avoid installing tools locally (like jdk) to work with the plugin.
I'm a Java developer since many years. If you want to work together, this will be a pleasure.

But I'm also agree with @Silicoman.
For me the first purpose is to be an official plugin by integration in official SonarQube Plugin repository. Thus, installation would be done by installing Jar files (with official SonarQube Plugin repository) and by a docker image.
For me docker image is only a tool to develop the plugin.

What do you think about it @jules-delecour-dav, @glalloue , @olegoaer, @mdubois81 ?

regards.

David.

**Comment by MP-Aubay** [comment]
Hi @obeone @dedece35 

I agree with @Silicoman, and for me the end goal is to automatically generate a plugin and adding it in Sonarqube Marketplace.

Docker should be use only for development and testing in our case.

Regards,
Mikaël

**Comment by dedece35** [comment]
Hi @obeone ,
we are migrating this repo to a new one : please see https://github.com/green-code-initiative/ecocode/pull/8
that's why I change this PR to "draft".

**Comment by Jberque** [comment]
Hi @obeone @dedece35 @MP-Aubay @Silicoman 
I just received a message from a "QA Test Manager" wishing to test Ecocode (Java), 
He sent me a message through Linkedin, telling me he's struggling to install the plugin.
If an all-in-one-docker-container exists it will be easier to use. 
I just wonder how many potential tester just drop the test because it was too hard to install instead of asking for support.
